src/samples/permissionhooks/AllowlistPlugin.sol

```diff
@@ -0,0 +1,142 @@
+        selectorAllowlist;
+
+    error TargetNotAllowed();
+    error SelectorNotAllowed();
+    error NoSelectorSpecified();
+
+    function onInstall(bytes calldata data) external override {
+        AllowlistInit[] memory init = abi.decode(data, (AllowlistInit[]));
+
+        for (uint256 i = 0; i < init.length; i++) {
+            targetAllowlist[init[i].target][msg.sender] = AllowlistEntry(true, init[i].hasSelectorAllowlist);
+
+            if (init[i].hasSelectorAllowlist) {
+                for (uint256 j = 0; j < init[i].selectors.length; j++) {
+                    selectorAllowlist[init[i].target][init[i].selectors[j]][msg.sender] = true;
+                }
+            }
+        }
+    }
+
+    function onUninstall(bytes calldata data) external override {
+        AllowlistInit[] memory init = abi.decode(data, (AllowlistInit[]));
+
+        for (uint256 i = 0; i < init.length; i++) {
+            delete targetAllowlist[init[i].target][msg.sender];
+
+            if (init[i].hasSelectorAllowlist) {
+                for (uint256 j = 0; j < init[i].selectors.length; j++) {
+                    delete selectorAllowlist[init[i].target][init[i].selectors[j]][msg.sender];
+                }
+            }
+        }
+    }
+
+    function setAllowlistTarget(address target, bool allowed, bool hasSelectorAllowlist) external {
+        targetAllowlist[target][msg.sender] = AllowlistEntry(allowed, hasSelectorAllowlist);
+    }
+
+    function setAllowlistSelector(address target, bytes4 selector, bool allowed) external {
+        selectorAllowlist[target][selector][msg.sender] = allowed;
+    }
+
+    function preUserOpValidationHook(uint8 functionId, PackedUserOperation calldata userOp, bytes32)
+        external
+        view
+        override
+        returns (uint256)
+    {
+        if (functionId == uint8(FunctionId.PRE_VALIDATION_HOOK)) {
+            _checkAllowlistCalldata(userOp.callData);
+            return 0;
+        }
+        revert NotImplemented();
+    }
+
+    function preRuntimeValidationHook(uint8 functionId, address, uint256, bytes calldata data, bytes calldata)
+        external
+        view
+        override
+    {
+        if (functionId == uint8(FunctionId.PRE_VALIDATION_HOOK)) {
+            _checkAllowlistCalldata(data);
+            return;
+        }
+
+        revert NotImplemented();
+    }
+
+    function pluginMetadata() external pure override returns (PluginMetadata memory) {
+        PluginMetadata memory metadata;
+        metadata.name = "Allowlist Plugin";
+        metadata.version = "v0.0.1";
+        metadata.author = "ERC-6900 Working Group";
+
+        return metadata;
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function pluginManifest() external pure override returns (PluginManifest memory) {}
+
+    function _checkAllowlistCalldata(bytes calldata callData) internal view {
+        if (bytes4(callData[:4]) == IStandardExecutor.execute.selector) {
```

Comment:
Callout: when executeUserOp is used, we need to check calldata[4:8] instead

Reply:
If this hook doesn't depend on the contents of `executeUserOp`, then the account shouldn't send the user op, right? I guess this depends on how we do it, but it seems like the account should know whether to use `msg.data` or `userOp.callData` depending on the hook config.

Reply:
yeah agree that it should be abstracted from plugins. in the current permissions stack, we directly send over userOp.callData which could be `abi.encodeCall(...)` or `abi.encodePacked(executeUserOp.selector, abi.encodeCall(...))` for hooks that don't require UO context, and for hooks that require UO context, they would always receive `abi.encodePacked(executeUserOp.selector, abi.encodeCall(...))`. Dislike that part and would likely change that though